specs/error-handling-guideline.md

```diff
@@ -0,0 +1,159 @@
+
+Last, error logs can also be useful for post-mortem debugging, truncate them occasionally so they don't eat up space on disk, and make sure they don't contain ansi color codes. Thereby, error logs can be written to a file.
+
+## Error output recommendation
+
+### Dos
+
+- Provide full description if the user input does not match what Notation CLI expected. A full description should include the actual input received from the user and expected input
+- Use the capital letter ahead of each line of any error message
+- Print human readable error message. If the error message is mainly from the server and varies by different servers, tell users that the error response is from server. This implies that users may need to contact server side for troubleshooting.
+- Provide specific and actionable prompt message with argument suggestion or show the example usage for reference. (e.g, Instead of showing flag or argument options is missing, please provide available argument options and guide users to "--help" to view more examples)
+- If the actionable prompt message is too long to show in the CLI output, consider guide users to Notation user guide or troubleshooting guide with the permanent link.
+- If the error message is not enough for troubleshooting, guide users to use "--verbose" to print much more detailed logs
+
+### Don'Ts
+
+- Do not use a formula-like or a programming expression in the error message. (e.g, `json: cannot unmarshal string into Go value of type map[string]map[string]string.`, or `Parameter 'xyz' must conform to the following pattern: '^[-\\w\\._\\(\\)]+$'`)
+- Do not use ambiguous expressions which mean nothing to users. (e.g, `Something unexpected happens`, or `Error: accepts 2 arg(s), received 0`)
+- Do not print irrelevant error message to make the output noisy. The more irrelevant output you produce, the longer it's going to take the user to figure out what they did wrong.
+
+## How to write friendly error message
+
+### Recommended error message structure
+
+Here is a sample structure of an error message:
+
+```text
+Error: [Error code]  [Error description] 
+Usage: [Command usage]
+[Recommended solution]
+```
+
+- Error code is an optional information. If the error message is generated from the server side, it may include error code or [warn code](https://www.rfc-editor.org/rfc/rfc7234#section-5.5). It could be printed out alongside the error description.
+- Command usage is also an optional information but it's recommended to be printed out when user input doesn't follow the standard usage or examples. 
+- Recommended solution is required and should follow the general guiding principles described above.
+
+### Examples
+
+#### Sign an artifact without artifact reference or signing key
+
+Current behavior and output:
```

Comment:
Btw, just a nit: 
I do find the word `Current` kind of awkward in this kind of scenario. Because it's semantically relative to any point in time of the reader, when we release a new version of Notation CLI, the behavior might then change making `Current behavior` no longer current anymore. I guess this pretty common in a guideline doc. Thus, my suggestion is, can we just remove these examples?

Reply:
I agree with moving all suggested changes for those examples into different GitHub issues. It would be much efficient for reviewing and managing this guidance content.